File: convert.py
import json
import subprocess
import os
import tempfile

# Function to test if Swift code compiles without errors
def test_swift_code(swift_code):
    with tempfile.NamedTemporaryFile(delete=False, suffix=".swift") as temp_file:
        temp_file_path = temp_file.name
        temp_file.write(swift_code.encode('utf-8'))
        temp_file.close()

        # Attempt to compile the Swift code
        # Modified to ignore encoding errors in stdout and stderr
        result = subprocess.run(['swift', temp_file_path], capture_output=True, text=True, errors='ignore')

        # Clean up the temporary file
        os.unlink(temp_file_path)

        # Return True if there's no error, False otherwise
        return result.stdout == "" and result.stderr == ""

# Prepare files to write supported and unsupported Swift code
supported_file = open('supported.swift', 'w', encoding='utf-8')
unsupported_file = open('unsupported.swift', 'w', encoding='utf-8')

# Read the JSON data
with open('data.raw.json', 'r', encoding='utf-8') as file:
    data = json.load(file)

for item in data:
    emoji = item['emoji']
    label = item['label'].replace(' ', '_')  # Replace spaces with underscores
    swift_line = f'let {emoji} = 1 /* {label} */\n'
    
    # Test if the Swift code compiles
    if test_swift_code(swift_line):
        supported_file.write(swift_line)
    else:
        unsupported_file.write(swift_line)

# Close the files
supported_file.close()
unsupported_file.close()

print('Done! Check supported.swift and unsupported.swift for the results.')

File: convert.swift
import Foundation

func testSwiftCode(swiftCode: String) -> Bool {
    let tempDirectory = NSTemporaryDirectory()
    let tempFileURL = URL(fileURLWithPath: tempDirectory).appendingPathComponent(UUID().uuidString).appendingPathExtension("swift")

    do {
        try swiftCode.write(to: tempFileURL, atomically: true, encoding: .utf8)
        
        let process = Process()
        let pipe = Pipe()
        process.executableURL = URL(fileURLWithPath: "/usr/bin/swift")
        process.arguments = [tempFileURL.path]
        process.standardOutput = pipe
        process.standardError = pipe
        
        try process.run()
        process.waitUntilExit()
        
        let data = pipe.fileHandleForReading.readDataToEndOfFile()
        let output = String(data: data, encoding: .utf8)
        
        try FileManager.default.removeItem(at: tempFileURL)
        
        return output?.isEmpty ?? false
    } catch {
        print("An error occurred: \(error.localizedDescription)")
        return false
    }
}
func main() {
    do {
        let fileManager = FileManager.default
        let currentDirectoryURL = URL(fileURLWithPath: fileManager.currentDirectoryPath)
        let jsonDataURL = currentDirectoryURL.appendingPathComponent("data.raw.json")
        
        // Change to write in the current directory
        let supportedFileURL = currentDirectoryURL.appendingPathComponent("supported.swift")
        let unsupportedFileURL = currentDirectoryURL.appendingPathComponent("unsupported.swift")
        
        let data = try Data(contentsOf: jsonDataURL)
        guard let jsonArray = try JSONSerialization.jsonObject(with: data, options: []) as? [[String: Any]] else {
            print("JSON data could not be parsed into the expected format.")
            return
        }
        
        var supportedSwiftCode = ""
        var unsupportedSwiftCode = ""
        
        for item in jsonArray {
            if let emoji = item["emoji"] as? String, let label = item["label"] as? String {
                 let swiftLine = "let \(emoji) = 1 /* \(label) */\n"
                
                if testSwiftCode(swiftCode: swiftLine) {
                    supportedSwiftCode += swiftLine
                } else {
                    unsupportedSwiftCode += swiftLine
                }
            }
        }
        
        try supportedSwiftCode.write(to: supportedFileURL, atomically: true, encoding: .utf8)
        try unsupportedSwiftCode.write(to: unsupportedFileURL, atomically: true, encoding: .utf8)
        
        print("Done! Check supported.swift and unsupported.swift for the results.")
    } catch {
        print("An error occurred: \(error.localizedDescription)")
    }
}
main()
